### PR TITLE
Feature/omemo

### PR DESCRIPTION
**Pull request by paulfariello**
Add native support to OMEMO

TODO:
- [x] OMEMO crypto materials generation
- [x] send OMEMO encrypted messages
- [x] receive OMEMO encrypted messages
- [x] long term storage for identity keys
- [x] add long term storage for session, ~~{signed,}prekeys,~~
- [x] get rid of libsodium
- [x] handle carbons
- [x] handle device_list_request response
- [x] check if prekey can be avoid while sending message
- [x] trust key from command line
- [x] untrust key from command line
- [x] valgrind check
- [ ] ~~add support for old gcm tag format parsing~~ (not being done for now since we are not aware of any client using it)
- [x] put gcm tag in key element (new format)
- [ ] ~~add support for signal pre v3 encryption~~ (not being done for now since we are not aware of any client using it)
- [x] start session on reception of a message with prekey (and republish bundle)
- [x] add support for libsignal 2.3.2
- [x] muc support
- [ ] ensure MUC has the right options (nonanonymous and access to fulljid)
- [x] Ensure PEP with bundle and device list are open (see [publish option](https://gist.github.com/iNPUTmice/7c52785ed69787516abb60e31703dbd2) )
- [ ] Check if `publish-options` is supported by server before attaching such node to query
- [x] Handle `precondition-not-met` error for `publish-options`
- [x] add autocompletion trust command
- [ ] add autocompletion untrust command
- [x] add a way to list trusted keys
- [ ] alert if sending message in a MUC with untrusted devices
- [ ] show list of known and trusted fingerprint on `/omemo start`
- [ ] add visual feedback when message has been sent on MUC

To test this PR you must install:

- gcrypt
- libsignal-protocol-c (>= v2.3.1)

On first run, ensure to create OMEMO cryptographic materials:
```
/omemo gen
```

To start an omemo session you must start it twice (issue about handling device_list_request response):
```
/omemo start buddy@example.org
```
Then you must trust some fingerprint:
```
/omemo fingerprint
/omemo trust <fingerprint>
```

### CODE REVIEW

**Comment by kaffeekanne** [comment]
To build this PR with profanity i had to remove  `-Werror` from `EM_CF_FLAGS` in `configure.ac`
i am no developer, how should it be actually built?

**Comment by paulfariello** [comment]
@kaffeekanne that's an issue. It should build without error. Can you provide a log of failing build?

**Comment by kaffeekanne** [comment]
Sure:
```
src/omemo/omemo.c: In function ‘omemo_init’:
src/omemo/omemo.c:108:30: error: initialization of ‘int (*)(signal_buffer **, signal_buffer **, const signal_protocol_address *, void *)’ {aka ‘int (*)(struct signal_buffer **, struct signal_buffer **, const struct signal_protocol_address *, void *)’} from incompatible pointer type ‘int (*)(signal_buffer **, const signal_protocol_address *, void *)’ {aka ‘int (*)(struct signal_buffer **, const struct signal_protocol_address *, void *)’} [-Werror=incompatible-pointer-types]

         .load_session_func = load_session,
                              ^~~~~~~~~~~~
src/omemo/omemo.c:108:30: note: (near initialization for ‘session_store.load_session_func’)
src/omemo/omemo.c:110:31: error: initialization of ‘int (*)(const signal_protocol_address *, uint8_t *, size_t,  uint8_t *, size_t,  void *)’ {aka ‘int (*)(const struct signal_protocol_address *, unsigned char *, long unsigned int,  unsigned char *, long unsigned int,  void *)’} from incompatible pointer type ‘int (*)(const signal_protocol_address *, uint8_t *, size_t,  void *)’int (*)(const struct signal_protocol_address *, unsigned char *, long unsigned int,  void *)’} [-Werror=incompatible-pointer-types]
         .store_session_func = store_session,
                               ^~~~~~~~~~~~~
src/omemo/omemo.c:110:31: note: (near initialization for ‘session_store.store_session_func’)
cc1: all warnings being treated as errors
make[1]: *** [Makefile:1823: src/omemo/omemo.o] Error 1
make[1]: Leaving directory '/home/ta/profanity'
make: *** [Makefile:1158: all] Error 2
```
Steps to reproduce:
```
git clone https://github.com/boothj5/profanity.git
cd profanity
git pull origin pull/1039/head
./bootstrap.sh
./configure
make
```
Do you set your build flags somewhere outside that configure.ac? Also i noticed `libsodium` is still being checked although you mention only needing `libgcrypt` and `libsignal-protocol-c`.

Also i `/omemo gen` does not give me any feedback and trying to start a omemo session does not work, because no crypto was generated. Any hints? I know this is alpha, but i am very interested in helping out.

**Comment by NiklausHofer** [comment]
@kaffeekanne I have observed the same behaviour with `/omemo gen`. You can work around this by restarting profanity after generating the keys.

**Comment by kaffeekanne** [comment]
@NiklausHofer ok, that did the trick for generation. Now i get the feedback, that everything has been created. Starting a OMEMO-Chat segfaults profanity.

**Comment by NiklausHofer** [comment]
@kaffeekanne I am getting this as well, yes. See my comment in the bug report: https://github.com/boothj5/profanity/issues/658#issuecomment-470662974

**Comment by paulfariello** [comment]
@kaffeekanne can you tell which version of libsignal you have? I suspect it's 2.3.2. Currently the PR only support 2.3.1.

**Comment by paulfariello** [comment]
Concerning the `/omemo gen` giving no feedback. I'll add this. And fix the issue of required restart after it.

**Comment by kaffeekanne** [comment]
`libsignal-protocol-c` is `2.3.2` using Archlinux. So every involved component should have the version of upstream.

**Comment by paulfariello** [comment]
Sadly libsignal broke the api between 2.3.1 and 2.3.2 😒

**Comment by kaffeekanne** [comment]
Building `libsignal-protocol-c` in version `2.3.1` results in warnings treated as errors as well. Maybe there is some setting on my system, that gets pulled. Anyway i like pleasing the compiler and getting rid of all warnings as well ;-)

**Comment by NiklausHofer** [comment]
I am using  2.3.1 (https://github.com/signalapp/libsignal-protocol-c/commit/aac6c6867ac6feb195a1d034da6e9cbb96959a69) though and as mentioned I am also getting the segfault.

**Comment by paulfariello** [comment]
@kaffeekanne just wait monday so I can add support for 2.3.2.

@niklaushofer yes your issue doesn't seems to be related to libsignal.

**Comment by paulfariello** [comment]
@kaffeekanne 61ec49695ccc8e5a98f1ee9902904cae74ef3a92 should add support for both 2.3.2 and 2.3.1

**Comment by paulfariello** [comment]
@NiklausHofer 35a11bfdac396848eb232b847650b8fbaeb1513c should fix your segfault.

**Comment by kaffeekanne** [comment]
I can compile with `libsignal-protocol-2=2.3.2` now and don't get segfaults. But i can't establish a OMEMO-session, did `/omemo start <jid>`, `/close`, `/omemo start <jid>`. The recepient is using gajim, if that matters. Anything i do to analyse? Where would you want to discuss such issues @paulfariello ?

**Comment by paulfariello** [comment]
You could give a look at logs and xmlconsole.
What happens when you send a message? The other end do not receive anything?
What happens when the other end tries to send you messages? You do not see anything?
I'm using gajim for some test and have no issue with it.

I'll have to add way more logs to debug this kind of issue.

**Comment by jubalh** [comment]
I think ` add support for libsignal 2.3.2` is already done, right? So this can be checked off? :)

**Comment by paulfariello** [comment]
@kaffeekanne last commits should fix the different issues you had. Ensure to remove `sessions.txt` in order to give libsignal-protocol-c a fresh start.

**Comment by kaffeekanne** [comment]
Started OMEMO-Sessions should be remembered, like it is done with gajim. If you have to `/omemo start` again always after profanity-restart, you easaly write unencrypted to e encrypted muc/contact by accident.

**Comment by pasis** [comment]
@paulfariello As #1025 points, Profanity can receive delayed OMEMO message (sent while offline). Do you handle such a situation in terms of keys or other state? I haven't read all the code yet, just asking to update your todo list if needed.

**Comment by paulfariello** [comment]
@pasis it should. But haven't tested it yet. Thanks for pointing that.

File: src/xmpp/omemo.c [hunk 1]
```diff
@@ -23,11 +23,9 @@ omemo_devicelist_publish(GList *device_list)
     xmpp_ctx_t * const ctx = connection_get_ctx();
     xmpp_stanza_t *iq = stanza_create_omemo_devicelist_publish(ctx, device_list);
 
-    Jid *jid = jid_create(connection_get_fulljid());
-    if (caps_jid_has_feature(jid->barejid, XMPP_FEATURE_PUBSUB_PUBLISH_OPTIONS)) {
+    if (connection_supports(XMPP_FEATURE_PUBSUB_PUBLISH_OPTIONS)) {
```
Comment (paulfariello) [inline src/xmpp/omemo.c]: `connection_supports` is fed with responses from `iq_disco_info_request_onconnect`.
But `omemo_devicelist_publish` is used from `omemo_on_connect` which is called in `sv_ev_login_account_success` so just before  `iq_disco_info_request_onconnect` see https://github.com/boothj5/profanity/blob/f7cb40ecf0b53a706875f1223dc99cf6481c3618/src/xmpp/session.c#L287-L320

@pasis @boothj5 any idea how this should be handled?

**Comment by jubalh** [comment]
@pasis @boothj5 I think @paulfariello is soon finishing his work on profanity. I think he did a great job and put in a lot of effort. There are no known bugs anymore and I would like to merge this soon.

I looked at most commits when they happened and think they are fine. We still need another round of codereview. I try to check it soon. It would be cool if you guys could also take a look @pasis, @boothj5 so we don't miss anything.

File: src/xmpp/session.c [hunk 2]
```diff
@@ -320,8 +320,8 @@ session_login_success(gboolean secured)
     blocking_request();
 
     // items discovery
+    connection_request_features();
```
Comment (paulfariello) [inline src/xmpp/session.c]: Not sure this is still required. The features are requested for main domain when we handle the disco items.

**Comment by jubalh** [comment]
Test:
/omemo trust <tab> to select a user
then remove the user jid and write your own (which is not in roster and thus not yet selectable via autocompletion), or another jid and hit <tab> again. It will not show/autocomplete the current keys.

We need to make sure:
* autocompletion completes for currently entered jid
* we need to be able to trust our own keys but don't list the one from the current device

**Comment by paulfariello** [comment]
> Test:
> /omemo trust to select a user
> then remove the user jid and write your own (which is not in roster and thus not yet selectable via autocompletion), or another jid and hit again. It will not show/autocomplete the current keys.
> 
> We need to make sure:
> 
>     * autocompletion completes for currently entered jid
> 
>     * we need to be able to trust our own keys but don't list the one from the current device

previous arguments are not passed to `autocomplete_func` in `autocomplete_param_no_with_func`. This is required to be able to filter fingerprint per jid.

**Comment by jubalh** [comment]
TODO:
OMEMO devicelist is only requested once `/omemo start jid` is run.

So the workflow:
```
/omemo gen
/omemo fingerprint examplejid
/omemo trust examplejid examplekey
```
does not work because fingerprint is empty.

If profantiy is build with OMEMO enabled we should request the devicelist on startup. So far this only happens in `/omemo start`.

**Comment by jubalh** [comment]
And we need to find out why always the first message that is sent is not encrypted.

**Comment by jubalh** [comment]
I think subtask#2 and subtask#3 should be fixed before a merge, preferably also subtask#1. Once we have this we can merge and track issues with real GitHub issues again. Since then basic functionality is working. And only features need to be added and maybe bugs fixed and UI improvements.

**Comment by kaffeekanne** [comment]
the work needed for subtask#1 are not special to `feature/omemo`. I noticed the need of rechecking already entered arguments before, though can not remember the situation. so i think that should be worked on another pr not related to omemo. Once that feature is available omemo and a lot of other stuff in profanity can make use of it and need a lot of fixing.

**Comment by jubalh** [comment]
subtask#2 and subtask#3 are solves with last commits.
Let's tackle subtask#1 in its own issue like suggested.

**Comment by jubalh** [comment]
Basic OMEMO functionality merged!

@paulfariello some of the tasks in the first comment are unchecked. You you reply down here which ones and why we chose to leave them out for now?

Maybe we can also create individual issues for the one that make sense.

**Comment by pasis** [comment]
Good job @paulfariello! @jubalh, sorry, I didn't have time to fully review the code. I will do it separately from the pull request.

**Pull request merged**